### PR TITLE
feat: add canGoBack

### PR DESCRIPTION
**Pull request by osdnk**

### CODE REVIEW

File: packages/core/src/useNavigationHelpers.tsx [hunk 1]
```diff
@@ -68,6 +68,9 @@ export default function useNavigationHelpers<Action extends NavigationAction>({
       isFocused: parentNavigationHelpers
         ? parentNavigationHelpers.isFocused
         : () => true,
+      canGoBack: parentNavigationHelpers
```
Comment (satya164) [inline packages/core/src/useNavigationHelpers.tsx]: maybe we also need to consider if current navigator can go back? it's different from `isFocused` cause `isFocused` checks if the navigator is focused, `canGoBack` checks if navigator can handle `goBack`.

so it could be something like:

```js
const canGoBack =
  router.canGoBack(getState()) ||
  parentNavigationHelpers
    ? parentNavigationHelpers.canGoBack()
    : false;
```

File: packages/core/src/useNavigationCache.tsx [hunk 2]
```diff
@@ -74,6 +83,13 @@ export default function useNavigationCache<
             return navigation ? navigation.isFocused() : true;
           },
           isFirstRouteInParent: () => isFirst,
+          canGoBack() {
```
Comment (satya164) [inline packages/core/src/useNavigationCache.tsx]: if we consider router state in the `canGoBack` of navigation helpers, we don't need to implement it here :)

Comment (satya164) [reply packages/core/src/useNavigationCache.tsx]: note: due to caching, this value will also be incorrect when state changes, so better to move it to navigation helpers

File: packages/core/src/useNavigationCache.tsx [hunk 3]
```diff
@@ -74,6 +83,13 @@ export default function useNavigationCache<
             return navigation ? navigation.isFocused() : true;
           },
           isFirstRouteInParent: () => isFirst,
```
Comment (satya164) [inline packages/core/src/useNavigationCache.tsx]: maybe we should remove this since we have `canGoBack`? not totally sure

File: packages/core/src/types.tsx [hunk 4]
```diff
@@ -154,6 +154,13 @@ export type Router<
    */
   shouldActionChangeFocus(action: NavigationAction): boolean;
 
+  /**
+   * Whether the back action will be handled by navigation
+   *
+   * @param state State object to apply the action on.
```
Comment (satya164) [inline packages/core/src/types.tsx]:
```suggestion
   * @param state State object to check.
```

File: packages/core/src/types.tsx [hunk 5]
```diff
@@ -264,6 +271,12 @@ type NavigationHelpersCommon<
    * To get notified of focus changes, use `addListener('focus', cb)` and `addListener('blur', cb)`.
    */
   isFocused(): boolean;
+
+  /**
+   * Check if dispatching back action will be handled by navigation.
+   * Note that this method doesn't re-render screen result changes. So don't u`se it in `render`.
```
Comment (satya164) [inline packages/core/src/types.tsx]: if we implement it like I mentioned, I think it'll work with re-renders :)

File: packages/core/src/BaseRouter.tsx [hunk 6]
```diff
@@ -78,6 +78,10 @@ const BaseRouter = {
   shouldActionChangeFocus(action: CommonAction) {
     return action.type === 'NAVIGATE';
   },
+
+  canGoBack() {
+    return false;
```
Comment (satya164) [inline packages/core/src/BaseRouter.tsx:83]: I think we shouldn't add it here. we should implement this property in tab router instead

File: packages/core/src/__tests__/__fixtures__/MockRouter.tsx [hunk 7]
```diff
@@ -98,6 +98,9 @@ export default function MockRouter(options: DefaultRouterOptions) {
     },
 
     actionCreators: {},
```
Comment (satya164) [inline packages/core/src/__tests__/__fixtures__/MockRouter.tsx]: nit: guess this `actionCreators` can be removed

File: packages/core/src/__tests__/useDescriptors.test.tsx [hunk 8]
```diff
@@ -54,6 +57,180 @@ it('sets options with options prop as an object', () => {
       `);
 });
 
+it('handles properly canGoBack with no canGoBack overrode', () => {
```
Comment (satya164) [inline packages/core/src/__tests__/useDescriptors.test.tsx]:
```suggestion
it("returns correct value for canGoBack when it's not overridden", () => {
```

File: packages/core/src/__tests__/useDescriptors.test.tsx [hunk 9]
```diff
@@ -54,6 +57,180 @@ it('sets options with options prop as an object', () => {
...
+      </TestNavigator>
+    </NavigationContainer>
+  );
+
+  render(root).update(root);
+
+  expect(result).toEqual(false);
+});
+
+it('handles properly canGoBack with canGoBack overrode', () => {
```
Comment (satya164) [inline packages/core/src/__tests__/useDescriptors.test.tsx]:
```suggestion
it("returns correct value for canGoBack when it's overridden", () => {
```

File: packages/core/src/__tests__/useDescriptors.test.tsx [hunk 10]
```diff
@@ -54,6 +57,180 @@ it('sets options with options prop as an object', () => {
...
+      </ParentNavigator>
+    </NavigationContainer>
+  );
+
+  render(root).update(root);
+
+  expect(result).toEqual(true);
+});
+
+it('handles properly canGoBack with canGoBack overrode but not as parent', () => {
```
Comment (satya164) [inline packages/core/src/__tests__/useDescriptors.test.tsx]:
```suggestion
it('returns correct value for canGoBack when parent router overrides it', () => {
```

File: packages/core/src/types.tsx (hunk 5 above)
Comment (satya164) [reply packages/core/src/types.tsx]: on another thought, maybe leave it. re-renders might trigger for a lot of things

File: packages/core/src/useNavigationCache.tsx (hunk 3 above)
Comment (osdnk) [reply packages/core/src/useNavigationCache.tsx]: not sure. Maybe in follow up

File: packages/core/src/BaseRouter.tsx (hunk 6 above)
Comment (osdnk) [reply packages/core/src/BaseRouter.tsx:83]: yes, but I'll still leave it by default for now

Comment (satya164) [reply packages/core/src/BaseRouter.tsx:83]: what I'm afraid of is that by leaving it here it's easy for routers to forget to implement it. since it's quite crucial to proper back button handling, maybe be it's better to force routers to implement it?

File: packages/core/src/NavigationContainer.tsx [hunk 11]
```diff
@@ -88,9 +88,7 @@ const Container = React.forwardRef(function NavigationContainer(
 
   const dispatch = (action: NavigationAction) =>
     context.performTransaction(() => {
-      for (let i = 0; i < actionListeners.length; i++) {
```
Comment (satya164) [inline packages/core/src/NavigationContainer.tsx:91]: why this change?

Comment (satya164) [reply packages/core/src/NavigationContainer.tsx:91]: oh I get it, cause there can only be one navigator, right?

File: packages/core/src/useNavigationBuilder.tsx [hunk 12]
```diff
@@ -198,6 +198,7 @@ export default function useNavigationBuilder<
     getState,
     setState,
     emitter,
+    router,
```
Comment (satya164) [inline packages/core/src/useNavigationBuilder.tsx:201]: unneeded?

File: packages/core/src/types.tsx (hunk 5 above)
Comment (satya164) [inline packages/core/src/types.tsx]:
```suggestion
   * Note that this method doesn't re-render screen when the result changes. So don't use it in `render`.
```

**Review by satya164 — APPROVED** [review]
LGTM, just few minor comments

File: packages/core/src/NavigationContainer.tsx (hunk 11 above)
Comment (satya164) [reply packages/core/src/NavigationContainer.tsx:91]: I'd keep the loop. the container shouldn't know that each navigator always adds only one listener

Comment (osdnk) [reply packages/core/src/NavigationContainer.tsx:91]: why? I see it needed bc we will need to return this value soon while wanted to implement back behavior

File: packages/core/src/useNavigationBuilder.tsx (hunk 12 above)
Comment (osdnk) [reply packages/core/src/useNavigationBuilder.tsx:201]: needed

Comment (osdnk) [reply packages/core/src/useNavigationBuilder.tsx:201]:
```
      canGoBack: () =>
        router.canGoBack(getState()) ||
        (parentNavigationHelpers && parentNavigationHelpers.canGoBack()) ||
        false,
```
